compiler/testData/diagnostics/tests/smartCasts/varnotnull/toFlexibleType.fir.kt

```diff
@@ -1,16 +1,25 @@
 // FILE: J.java
+import org.jetbrains.annotations.*;
+import java.util.List;
 
 class J {
     static String foo() { return "abc"; }
+
+    @NotNull
+    static List<String> bar() { return new List<String>(); }
 }
 
 // FILE: test.kt
 
 fun bar() {
+    var x: List<String>?
+    x = J.bar()
+    x.isEmpty()
+
     var v: String?
     v = J.foo()
-    v.length
-    gav(v)
+    v<!UNSAFE_CALL!>.<!>length
```

Comment:
Actually there must be smartcast, because `(String, String?)` is acceptable as non-nullable receiver

Reply:
`J.foo()` from Java, without any annotations, returns `ft(String..String?)`. So here `it(ft(String..String?), String?)` = `String?`.

Reply:
I think `it(ft(String..String?), String?)` should actually be `ft(String..String?)` since `ft(String..String?) <: String?`.

Reply:
Oh, so @demiurg906's point was, intersection result should be `ft(String..String?)`, and that's acceptable as non-nullable receiver? Let me take a look at this further.

Reply:
`ConeTypeIntersector` will filter out one participating type if that is a strict super type of others. That part checks if `X <: Y` is true and `Y <: X` is false. I expected `String?` needs to be filtered out, but it didn't, because both queries of `ft(String..String?) <: String?` and `String? <: ft(String..String?)` return true. The latter seems not right, and the root cause is here: https://github.com/JetBrains/kotlin/blob/master/core/compiler.common/src/org/jetbrains/kotlin/types/AbstractTypeChecker.kt#L249, which checks lower bound of sub type (`String?` in this case) and upper bound of super type (and `String?` again in this case). I wonder that should be the other way around: checking upper bound of sub type against lower bound of super type?

Touching that place in `AbstractTypeChecker` caused tons of regressions in FE 1.0, which I can't fully follow. Even some are simply broken (not just test data diff), and it looks like some assumptions in FE 1.0's flexible type is violated in many cases.

As a stopgap, in 3rd commit, I tweak those queries in `ConeTypeIntersector` to use upper bound of LHS and lower bound of RHS. With that, these smartcast changes are back to the previous states, while fixing incorrect nullability after type intersect, inspired by #4257 and #4265, still remain (i.e., `it(ft(J..J?), ft(J..J?)) = ft(J..J?)`).

Reply:
I found that https://github.com/JetBrains/kotlin/blob/master/core/compiler.common/src/org/jetbrains/kotlin/types/AbstractTypeChecker.kt#L249 is based on https://kotlinlang.org/spec/type-system.html#subtyping-for-flexible-types
```
L <: B  ⇒ (L..U) <: (A..B)
```
But, given L <: U in (L..U) and A <: B in (A..B), I don't understand how this implication holds. Should it be:
```
U <: A ⇒ (L..U) <: (A..B)
```
?

Same for other implications in the doc:
```
L <: T ⇒ (L..U) <: T
T <: U ⇒ T <: (L..U)
```

Given L <: U in (L..U), rather, what make sense to me are:
```
U <: T => (L..U) <: T
T <: L => T <: (L..U)
```
?

Reply:
I think the spec and the current subtype checker's behavior is correct. Flexible type is weird. It opens a hole in the Kotlin type theory and makes it unsound. I guess an intuitive way to understand flexible type `L..U` is: if there is some `T` with `L <: T <: U`, such that `P(T)` is true, then `P(L..U)` is true.

The proposition `L <: B  ⇒ (L..U) <: (A..B)` holds because one can choose `L` from `L..U` and and `B` from `A..B` to make the proposition hold.

So it sounds like the filtering done in the type intersector should handle flexible type specially: it should prefer flexible types (even finding union of multiple flexible types if needed) if subtyping relation deems it to be "equivalent" to another type.

Reply:
Too bad to see how the inter-operability can break the soundness of type system. All of those still look counter-intuitive to me. :'(

Anyway, I found another way to make `ConeTypeIntersector` pick flexible type over regular, other equivalent type, which corrects the smartcast here while preserving other good changes.

Reply:
Yep, flexible types are hole in type system and it is by design